File: src/features/quiz/store/quiz-store.ts
import { create } from "zustand";
import { Answer } from "@/features/quiz/types/answer";
import { Recommendation } from "@/features/recommendations/types/recommendation";

export type ContentType = "movie" | "book" | "music" | "both" | "mix";

interface QuizStore {
  contentType: ContentType | null;
  setContentType: (type: ContentType) => void;

  questionCount: number;
  setQuestionCount: (count: number) => void;

  answers: Answer[];
  setAnswers: (answers: Answer[]) => void;

  userAge: number | null;
  setUserAge: (age: number) => void;

  filters: { genres: string[]; moods: string[] };
  setFilters: (filters: { genres: string[]; moods: string[] }) => void;

  // Recommendations are pre-generated inside /quiz's generating step and
  // handed off here so /results can render them immediately on mount instead
  // of running its own loader. The store is in-memory only, so a hard reload
  // of /results clears these and the page falls back to its own gen flow.
  recommendations: Recommendation[];
  setRecommendations: (recs: Recommendation[]) => void;

  reset: () => void;
}

const defaultState = {
  contentType: null as ContentType | null,
  questionCount: 5,
  answers: [] as Answer[],
  userAge: null as number | null,
  filters: { genres: [] as string[], moods: [] as string[] },
  recommendations: [] as Recommendation[],
};

export const useQuizStore = create<QuizStore>((set) => ({
  ...defaultState,

  setContentType: (contentType) => set({ contentType }),
  setQuestionCount: (questionCount) => set({ questionCount }),
  setAnswers: (answers) => set({ answers }),
  setUserAge: (userAge) => set({ userAge }),
  setFilters: (filters) => set({ filters }),
  setRecommendations: (recommendations) => set({ recommendations }),
  reset: () => set(defaultState),
}));

File: apps/mobile/lib/features/quiz/store/quiz_store.dart
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:freezed_annotation/freezed_annotation.dart';

import '../../../core/models/answer.dart';
import '../../../core/models/enums.dart';
import '../../../core/models/recommendation.dart';

part 'quiz_store.freezed.dart';

/// Port of web quiz-store.ts (Zustand). In-memory only — a hard reload
/// clears it, matching web behavior.
@freezed
abstract class QuizState with _$QuizState {
  const factory QuizState({
    ContentType? contentType,
    @Default(5) int questionCount,
    @Default(<Answer>[]) List<Answer> answers,
    int? userAge,
    @Default(<String>[]) List<String> genres,
    @Default(<String>[]) List<String> moods,
    @Default(<Recommendation>[]) List<Recommendation> recommendations,
  }) = _QuizState;
}

class QuizStore extends Notifier<QuizState> {
  @override
  QuizState build() => const QuizState();

  void setContentType(ContentType type) =>
      state = state.copyWith(contentType: type);

  void setQuestionCount(int count) =>
      state = state.copyWith(questionCount: count);

  void setAnswers(List<Answer> answers) =>
      state = state.copyWith(answers: answers);

  void setUserAge(int age) => state = state.copyWith(userAge: age);

  void setFilters({required List<String> genres, required List<String> moods}) =>
      state = state.copyWith(genres: genres, moods: moods);

  void setRecommendations(List<Recommendation> recs) =>
      state = state.copyWith(recommendations: recs);

  void reset() => state = const QuizState();
}

final quizStoreProvider =
    NotifierProvider<QuizStore, QuizState>(QuizStore.new);
